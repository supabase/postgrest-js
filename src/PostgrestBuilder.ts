import crossFetch from 'cross-fetch'

import type { Fetch, PostgrestResponse } from './types'

export default abstract class PostgrestBuilder<Result>
  implements PromiseLike<PostgrestResponse<Result>>
{
  protected method: 'GET' | 'HEAD' | 'POST' | 'PATCH' | 'DELETE'
  protected url: URL
  protected headers: Record<string, string>
  protected schema?: string
  protected body?: unknown
  protected shouldThrowOnError = false
  protected signal?: AbortSignal
  protected fetch: Fetch
  protected allowEmpty: boolean

  constructor(builder: PostgrestBuilder<Result>) {
    this.method = builder.method
    this.url = builder.url
    this.headers = builder.headers
    this.schema = builder.schema
    this.body = builder.body
    this.shouldThrowOnError = builder.shouldThrowOnError
    this.signal = builder.signal
    this.allowEmpty = builder.allowEmpty

    if (builder.fetch) {
      this.fetch = builder.fetch
    } else if (typeof fetch === 'undefined') {
      this.fetch = crossFetch
    } else {
      this.fetch = fetch
    }
  }

  /**
   * If there's an error with the query, throwOnError will reject the promise by
   * throwing the error instead of returning it as part of a successful response.
   *
   * {@link https://github.com/supabase/supabase-js/issues/92}
   */
  throwOnError(): this {
    this.shouldThrowOnError = true
    return this
  }

  then<TResult1 = PostgrestResponse<Result>, TResult2 = never>(
    onfulfilled?:
      | ((value: PostgrestResponse<Result>) => TResult1 | PromiseLike<TResult1>)
      | undefined
      | null,
    onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null
  ): PromiseLike<TResult1 | TResult2> {
    // https://postgrest.org/en/stable/api.html#switching-schemas
    if (this.schema === undefined) {
      // skip
    } else if (['GET', 'HEAD'].includes(this.method)) {
      this.headers['Accept-Profile'] = this.schema
    } else {
      this.headers['Content-Profile'] = this.schema
    }
    if (this.method !== 'GET' && this.method !== 'HEAD') {
      this.headers['Content-Type'] = 'application/json'
    }

    // NOTE: Invoke w/o `this` to avoid illegal invocation error.
    // https://github.com/supabase/postgrest-js/pull/247
    const _fetch = this.fetch
    let res = _fetch(this.url.toString(), {
      method: this.method,
      headers: this.headers,
      body: JSON.stringify(this.body),
      signal: this.signal,
    }).then(async (res) => {
      let error = null
      let data = null
      let count: number | null = null
      let status = res.status
      let statusText = res.statusText

      if (res.ok) {
        if (this.method !== 'HEAD') {
          const body = await res.text()
          if (body === '') {
            // Prefer: return=minimal
          } else if (this.headers['Accept'] === 'text/csv') {
            data = body
          } else if (
            this.headers['Accept'] &&
            this.headers['Accept'].includes('application/vnd.pgrst.plan+text')
          ) {
            data = body
          } else {
            data = JSON.parse(body)
          }
        }

        const countHeader = this.headers['Prefer']?.match(/count=(exact|planned|estimated)/)
        const contentRange = res.headers.get('content-range')?.split('/')
        if (countHeader && contentRange && contentRange.length > 1) {
          count = parseInt(contentRange[1])
        }
      } else {
        const body = await res.text()

        try {
          error = JSON.parse(body)
        } catch {
          error = {
            message: body,
          }
        }

        if (error && this.allowEmpty && error?.details?.includes('Results contain 0 rows')) {
          error = null
          status = 200
          statusText = 'OK'
        }

        if (error && this.shouldThrowOnError) {
          throw error
        }
      }

      const postgrestResponse = {
        error,
        data,
        count,
        status,
        statusText,
      }

      return postgrestResponse
    })
    if (!this.shouldThrowOnError) {
      res = res.catch((fetchError) => ({
        error: {
          message: `FetchError: ${fetchError.message}`,
          details: '',
          hint: '',
          code: fetchError.code || '',
        },
        data: null,
        count: null,
        status: 0,
        statusText: '',
      }))
    }

    return res.then(onfulfilled, onrejected)
  }
}
