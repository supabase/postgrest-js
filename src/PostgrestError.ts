import type { PostgrestError as IPostgrestError } from './types'

export default class PostgrestError extends Error implements IPostgrestError {
  details?: string
  hint?: string
  code?: string

  constructor(context: IPostgrestError) {
    super(context.message)
    this.name = 'PostgrestError'
    this.details = context.details
    this.hint = context.hint
    this.code = context.code
  }
}
