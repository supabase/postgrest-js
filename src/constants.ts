import { version } from './version'
export const DEFAULT_HEADERS = { 'X-Client-Info': `postgrest-js/${version}` }
