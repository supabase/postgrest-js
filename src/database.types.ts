export type Json = string | number | boolean | null | { [key: string]: Json | undefined } | Json[]

export type Database = {
  graphql_public: {
    Tables: {
      [_ in never]: never
    }
    Views: {
      [_ in never]: never
    }
    Functions: {
      graphql: {
        Args: {
          operationName?: string
          query?: string
          variables?: Json
          extensions?: Json
        }
        Returns: Json
      }
    }
    Enums: {
      [_ in never]: never
    }
    CompositeTypes: {
      [_ in never]: never
    }
  }
  public: {
    Tables: {
      files: {
        Row: {
          created_at: string
          id: number
          name: string
          owner_id: string
          path: string
          size: number
          type: string | null
          updated_at: string
        }
        Insert: {
          created_at?: string
          id?: number
          name: string
          owner_id: string
          path: string
          size: number
          type?: string | null
          updated_at?: string
        }
        Update: {
          created_at?: string
          id?: number
          name?: string
          owner_id?: string
          path?: string
          size?: number
          type?: string | null
          updated_at?: string
        }
        Relationships: [
          {
            foreignKeyName: 'files_owner_id_fkey'
            columns: ['owner_id']
            isOneToOne: false
            referencedRelation: 'user_share_info'
            referencedColumns: ['id']
          },
          {
            foreignKeyName: 'files_owner_id_fkey'
            columns: ['owner_id']
            isOneToOne: false
            referencedRelation: 'users'
            referencedColumns: ['id']
          }
        ]
      }
      shared_files: {
        Row: {
          created_at: string
          file_id: number
          id: number
          shared_by: string
          shared_with: string
          updated_at: string
        }
        Insert: {
          created_at?: string
          file_id: number
          id?: number
          shared_by: string
          shared_with?: string | null
          updated_at?: string
        }
        Update: {
          created_at?: string
          file_id?: number
          id?: number
          shared_by?: string
          shared_with?: string | null
          updated_at?: string
        }
        Relationships: [
          {
            foreignKeyName: 'shared_files_file_id_fkey'
            columns: ['file_id']
            isOneToOne: false
            referencedRelation: 'files'
            referencedColumns: ['id']
          },
          {
            foreignKeyName: 'shared_files_shared_by_fkey'
            columns: ['shared_by']
            isOneToOne: false
            referencedRelation: 'user_share_info'
            referencedColumns: ['id']
          },
          {
            foreignKeyName: 'shared_files_shared_by_fkey'
            columns: ['shared_by']
            isOneToOne: false
            referencedRelation: 'users'
            referencedColumns: ['id']
          },
          {
            foreignKeyName: 'shared_files_shared_with_fkey'
            columns: ['shared_with']
            isOneToOne: false
            referencedRelation: 'user_share_info'
            referencedColumns: ['id']
          },
          {
            foreignKeyName: 'shared_files_shared_with_fkey'
            columns: ['shared_with']
            isOneToOne: false
            referencedRelation: 'users'
            referencedColumns: ['id']
          }
        ]
      }
    }
    Views: {
      user_share_info: {
        Row: {
          avatar_url: string | null
          email: string | null
          id: string | null
          name: string | null
        }
        Insert: {
          avatar_url?: never
          email?: string | null
          id?: string | null
          name?: never
        }
        Update: {
          avatar_url?: never
          email?: string | null
          id?: string | null
          name?: never
        }
        Relationships: []
      }
    }
    Functions: {
      is_file_shared_with_user: {
        Args: {
          file_path: string
        }
        Returns: boolean
      }
    }
    Enums: {
      [_ in never]: never
    }
    CompositeTypes: {
      [_ in never]: never
    }
  }
}

type PublicSchema = Database[Extract<keyof Database, 'public'>]

export type Tables<
  PublicTableNameOrOptions extends
    | keyof (PublicSchema['Tables'] & PublicSchema['Views'])
    | { schema: keyof Database },
  TableName extends PublicTableNameOrOptions extends { schema: keyof Database }
    ? keyof (Database[PublicTableNameOrOptions['schema']]['Tables'] &
        Database[PublicTableNameOrOptions['schema']]['Views'])
    : never = never
> = PublicTableNameOrOptions extends { schema: keyof Database }
  ? (Database[PublicTableNameOrOptions['schema']]['Tables'] &
      Database[PublicTableNameOrOptions['schema']]['Views'])[TableName] extends {
      Row: infer R
    }
    ? R
    : never
  : PublicTableNameOrOptions extends keyof (PublicSchema['Tables'] & PublicSchema['Views'])
  ? (PublicSchema['Tables'] & PublicSchema['Views'])[PublicTableNameOrOptions] extends {
      Row: infer R
    }
    ? R
    : never
  : never

export type TablesInsert<
  PublicTableNameOrOptions extends keyof PublicSchema['Tables'] | { schema: keyof Database },
  TableName extends PublicTableNameOrOptions extends { schema: keyof Database }
    ? keyof Database[PublicTableNameOrOptions['schema']]['Tables']
    : never = never
> = PublicTableNameOrOptions extends { schema: keyof Database }
  ? Database[PublicTableNameOrOptions['schema']]['Tables'][TableName] extends {
      Insert: infer I
    }
    ? I
    : never
  : PublicTableNameOrOptions extends keyof PublicSchema['Tables']
  ? PublicSchema['Tables'][PublicTableNameOrOptions] extends {
      Insert: infer I
    }
    ? I
    : never
  : never

export type TablesUpdate<
  PublicTableNameOrOptions extends keyof PublicSchema['Tables'] | { schema: keyof Database },
  TableName extends PublicTableNameOrOptions extends { schema: keyof Database }
    ? keyof Database[PublicTableNameOrOptions['schema']]['Tables']
    : never = never
> = PublicTableNameOrOptions extends { schema: keyof Database }
  ? Database[PublicTableNameOrOptions['schema']]['Tables'][TableName] extends {
      Update: infer U
    }
    ? U
    : never
  : PublicTableNameOrOptions extends keyof PublicSchema['Tables']
  ? PublicSchema['Tables'][PublicTableNameOrOptions] extends {
      Update: infer U
    }
    ? U
    : never
  : never

export type Enums<
  PublicEnumNameOrOptions extends keyof PublicSchema['Enums'] | { schema: keyof Database },
  EnumName extends PublicEnumNameOrOptions extends { schema: keyof Database }
    ? keyof Database[PublicEnumNameOrOptions['schema']]['Enums']
    : never = never
> = PublicEnumNameOrOptions extends { schema: keyof Database }
  ? Database[PublicEnumNameOrOptions['schema']]['Enums'][EnumName]
  : PublicEnumNameOrOptions extends keyof PublicSchema['Enums']
  ? PublicSchema['Enums'][PublicEnumNameOrOptions]
  : never
