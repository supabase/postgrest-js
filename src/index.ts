export { default as PostgrestClient } from './PostgrestClient'
export { default as PostgrestQueryBuilder } from './PostgrestQueryBuilder'
export { default as PostgrestFilterBuilder } from './PostgrestFilterBuilder'
export { default as PostgrestTransformBuilder } from './PostgrestTransformBuilder'
export { default as PostgrestBuilder } from './PostgrestBuilder'
export {
  PostgrestResponse,
  PostgrestSingleResponse,
  PostgrestMaybeSingleResponse,
  PostgrestError,
} from './types'
