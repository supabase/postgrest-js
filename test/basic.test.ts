import { PostgrestClient } from '../src/index'

const REST_URL = 'http://localhost:3000'
const postgrest = new PostgrestClient(REST_URL)

test('basic select table', async () => {
  const res = await postgrest.from('users').select()
  expect(res).toMatchSnapshot()
})

test('stored procedure', async () => {
  const res = await postgrest.rpc('get_status', { name_param: 'supabot' })
  expect(res).toMatchSnapshot()
})

test('custom headers', async () => {
  const postgrest = new PostgrestClient(REST_URL, { headers: { apikey: 'foo' } })
  expect((postgrest.from('users').select() as any).headers['apikey']).toEqual('foo')
})

test('auth', async () => {
  const postgrest = new PostgrestClient(REST_URL).auth('foo')
  expect((postgrest.from('users').select() as any).headers['Authorization']).toEqual('Bearer foo')
})

test('switch schema', async () => {
  const postgrest = new PostgrestClient(REST_URL, { schema: 'personal' })
  const res = await postgrest.from('users').select()
  expect(res).toMatchSnapshot()
})

test('on_conflict insert', async () => {
  const res = await postgrest
    .from('users')
    .insert({ username: 'dragarcia' }, { upsert: true, onConflict: 'username' })
  expect(res).toMatchSnapshot()
})

describe('basic insert, update, delete', () => {
  test('basic insert', async () => {
    let res = await postgrest
      .from('messages')
      .insert({ message: 'foo', username: 'supabot', channel_id: 1 })
    expect(res).toMatchSnapshot()

    res = await postgrest.from('messages').select()
    expect(res).toMatchSnapshot()
  })

  test('upsert', async () => {
    let res = await postgrest
      .from('messages')
      .insert({ id: 3, message: 'foo', username: 'supabot', channel_id: 2 }, { upsert: true })
    expect(res).toMatchSnapshot()

    res = await postgrest.from('messages').select()
    expect(res).toMatchSnapshot()
  })

  test('bulk insert', async () => {
    let res = await postgrest.from('messages').insert([
      { message: 'foo', username: 'supabot', channel_id: 1 },
      { message: 'foo', username: 'supabot', channel_id: 1 },
    ])
    expect(res).toMatchSnapshot()

    res = await postgrest.from('messages').select()
    expect(res).toMatchSnapshot()
  })

  test('basic update', async () => {
    let res = await postgrest.from('messages').update({ channel_id: 2 }).eq('message', 'foo')
    expect(res).toMatchSnapshot()

    res = await postgrest.from('messages').select()
    expect(res).toMatchSnapshot()
  })

  test('basic delete', async () => {
    let res = await postgrest.from('messages').delete().eq('message', 'foo')
    expect(res).toMatchSnapshot()

    res = await postgrest.from('messages').select()
    expect(res).toMatchSnapshot()
  })
})

test('missing table', async () => {
  const res = await postgrest.from('missing_table').select()
  expect(res).toMatchSnapshot()
})

test('connection error', async () => {
  const postgrest = new PostgrestClient('http://this.url.does.not.exist')
  let isErrorCaught = false
  await postgrest
    .from('user')
    .select()
    .then(undefined, () => {
      isErrorCaught = true
    })
  expect(isErrorCaught).toBe(true)
})

test('custom type', async () => {
  interface User {
    username: string
    data: object | null
    age_range: string | null
    status: 'ONLINE' | 'OFFLINE'
    catchphrase: 'string' | null
  }

  // TODO: Find a cleaner way to weave a custom type
  // eq should show User's properties in LSP/IntelliSense
  const { data: users } = <{ data: User[] }>(
    await postgrest.from<User>('users').select().eq('username', 'supabot')
  )
  const user = users[0]
  // Autocomplete should show properties of user after '.'
  user.username
})

test("don't mutate PostgrestClient.headers", async () => {
  await postgrest.from('users').select().limit(1).single()
  const { error } = await postgrest.from('users').select()
  expect(error).toMatchSnapshot()
})

test('allow ordering on JSON column', async () => {
  const { data } = await postgrest.from('users').select().order('data->something')
  expect(data).toMatchSnapshot()
})

test('Prefer: return=minimal', async () => {
  const { data } = await postgrest.from('users').insert({ username: 'bar' }, { returning: 'minimal' })
  expect(data).toMatchSnapshot()

  await postgrest.from('users').delete().eq('username', 'bar')
})
