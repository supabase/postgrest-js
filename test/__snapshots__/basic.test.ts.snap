// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prefer: return=minimal 1`] = `null`;

exports[`allow ordering on JSON column 1`] = `
Array [
  Object {
    "age_range": "[1,2)",
    "catchphrase": "'cat' 'fat'",
    "data": null,
    "status": "ONLINE",
    "username": "supabot",
  },
  Object {
    "age_range": "[25,35)",
    "catchphrase": "'bat' 'cat'",
    "data": null,
    "status": "OFFLINE",
    "username": "kiwicopple",
  },
  Object {
    "age_range": "[25,35)",
    "catchphrase": "'bat' 'rat'",
    "data": null,
    "status": "ONLINE",
    "username": "awailas",
  },
  Object {
    "age_range": "[20,30)",
    "catchphrase": "'fat' 'rat'",
    "data": null,
    "status": "ONLINE",
    "username": "dragarcia",
  },
]
`;

exports[`basic insert, update, delete basic delete 1`] = `
Object {
  "body": Array [
    Object {
      "channel_id": 2,
      "data": null,
      "id": 3,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 4,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 5,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "data": Array [
    Object {
      "channel_id": 2,
      "data": null,
      "id": 3,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 4,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 5,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "error": null,
  "status": 200,
  "statusText": "OK",
}
`;

exports[`basic insert, update, delete basic delete 2`] = `
Object {
  "body": Array [
    Object {
      "channel_id": 1,
      "data": null,
      "id": 1,
      "message": "Hello World ðŸ‘‹",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 2,
      "message": "Perfection is attained, not when there is nothing more to add, but when there is nothing left to take away.",
      "username": "supabot",
    },
  ],
  "data": Array [
    Object {
      "channel_id": 1,
      "data": null,
      "id": 1,
      "message": "Hello World ðŸ‘‹",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 2,
      "message": "Perfection is attained, not when there is nothing more to add, but when there is nothing left to take away.",
      "username": "supabot",
    },
  ],
  "error": null,
  "status": 200,
  "statusText": "OK",
}
`;

exports[`basic insert, update, delete basic insert 1`] = `
Object {
  "body": Array [
    Object {
      "channel_id": 1,
      "data": null,
      "id": 3,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "data": Array [
    Object {
      "channel_id": 1,
      "data": null,
      "id": 3,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "error": null,
  "status": 201,
  "statusText": "Created",
}
`;

exports[`basic insert, update, delete basic insert 2`] = `
Object {
  "body": Array [
    Object {
      "channel_id": 1,
      "data": null,
      "id": 1,
      "message": "Hello World ðŸ‘‹",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 2,
      "message": "Perfection is attained, not when there is nothing more to add, but when there is nothing left to take away.",
      "username": "supabot",
    },
    Object {
      "channel_id": 1,
      "data": null,
      "id": 3,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "data": Array [
    Object {
      "channel_id": 1,
      "data": null,
      "id": 1,
      "message": "Hello World ðŸ‘‹",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 2,
      "message": "Perfection is attained, not when there is nothing more to add, but when there is nothing left to take away.",
      "username": "supabot",
    },
    Object {
      "channel_id": 1,
      "data": null,
      "id": 3,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "error": null,
  "status": 200,
  "statusText": "OK",
}
`;

exports[`basic insert, update, delete basic update 1`] = `
Object {
  "body": Array [
    Object {
      "channel_id": 2,
      "data": null,
      "id": 3,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 4,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 5,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "data": Array [
    Object {
      "channel_id": 2,
      "data": null,
      "id": 3,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 4,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 5,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "error": null,
  "status": 200,
  "statusText": "OK",
}
`;

exports[`basic insert, update, delete basic update 2`] = `
Object {
  "body": Array [
    Object {
      "channel_id": 1,
      "data": null,
      "id": 1,
      "message": "Hello World ðŸ‘‹",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 2,
      "message": "Perfection is attained, not when there is nothing more to add, but when there is nothing left to take away.",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 3,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 4,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 5,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "data": Array [
    Object {
      "channel_id": 1,
      "data": null,
      "id": 1,
      "message": "Hello World ðŸ‘‹",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 2,
      "message": "Perfection is attained, not when there is nothing more to add, but when there is nothing left to take away.",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 3,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 4,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 5,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "error": null,
  "status": 200,
  "statusText": "OK",
}
`;

exports[`basic insert, update, delete bulk insert 1`] = `
Object {
  "body": Array [
    Object {
      "channel_id": 1,
      "data": null,
      "id": 4,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 1,
      "data": null,
      "id": 5,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "data": Array [
    Object {
      "channel_id": 1,
      "data": null,
      "id": 4,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 1,
      "data": null,
      "id": 5,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "error": null,
  "status": 201,
  "statusText": "Created",
}
`;

exports[`basic insert, update, delete bulk insert 2`] = `
Object {
  "body": Array [
    Object {
      "channel_id": 1,
      "data": null,
      "id": 1,
      "message": "Hello World ðŸ‘‹",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 2,
      "message": "Perfection is attained, not when there is nothing more to add, but when there is nothing left to take away.",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 3,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 1,
      "data": null,
      "id": 4,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 1,
      "data": null,
      "id": 5,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "data": Array [
    Object {
      "channel_id": 1,
      "data": null,
      "id": 1,
      "message": "Hello World ðŸ‘‹",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 2,
      "message": "Perfection is attained, not when there is nothing more to add, but when there is nothing left to take away.",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 3,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 1,
      "data": null,
      "id": 4,
      "message": "foo",
      "username": "supabot",
    },
    Object {
      "channel_id": 1,
      "data": null,
      "id": 5,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "error": null,
  "status": 200,
  "statusText": "OK",
}
`;

exports[`basic insert, update, delete upsert 1`] = `
Object {
  "body": Array [
    Object {
      "channel_id": 2,
      "data": null,
      "id": 3,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "data": Array [
    Object {
      "channel_id": 2,
      "data": null,
      "id": 3,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "error": null,
  "status": 201,
  "statusText": "Created",
}
`;

exports[`basic insert, update, delete upsert 2`] = `
Object {
  "body": Array [
    Object {
      "channel_id": 1,
      "data": null,
      "id": 1,
      "message": "Hello World ðŸ‘‹",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 2,
      "message": "Perfection is attained, not when there is nothing more to add, but when there is nothing left to take away.",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 3,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "data": Array [
    Object {
      "channel_id": 1,
      "data": null,
      "id": 1,
      "message": "Hello World ðŸ‘‹",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 2,
      "message": "Perfection is attained, not when there is nothing more to add, but when there is nothing left to take away.",
      "username": "supabot",
    },
    Object {
      "channel_id": 2,
      "data": null,
      "id": 3,
      "message": "foo",
      "username": "supabot",
    },
  ],
  "error": null,
  "status": 200,
  "statusText": "OK",
}
`;

exports[`basic select table 1`] = `
Object {
  "body": Array [
    Object {
      "age_range": "[1,2)",
      "catchphrase": "'cat' 'fat'",
      "data": null,
      "status": "ONLINE",
      "username": "supabot",
    },
    Object {
      "age_range": "[25,35)",
      "catchphrase": "'bat' 'cat'",
      "data": null,
      "status": "OFFLINE",
      "username": "kiwicopple",
    },
    Object {
      "age_range": "[25,35)",
      "catchphrase": "'bat' 'rat'",
      "data": null,
      "status": "ONLINE",
      "username": "awailas",
    },
    Object {
      "age_range": "[20,30)",
      "catchphrase": "'fat' 'rat'",
      "data": null,
      "status": "ONLINE",
      "username": "dragarcia",
    },
  ],
  "data": Array [
    Object {
      "age_range": "[1,2)",
      "catchphrase": "'cat' 'fat'",
      "data": null,
      "status": "ONLINE",
      "username": "supabot",
    },
    Object {
      "age_range": "[25,35)",
      "catchphrase": "'bat' 'cat'",
      "data": null,
      "status": "OFFLINE",
      "username": "kiwicopple",
    },
    Object {
      "age_range": "[25,35)",
      "catchphrase": "'bat' 'rat'",
      "data": null,
      "status": "ONLINE",
      "username": "awailas",
    },
    Object {
      "age_range": "[20,30)",
      "catchphrase": "'fat' 'rat'",
      "data": null,
      "status": "ONLINE",
      "username": "dragarcia",
    },
  ],
  "error": null,
  "status": 200,
  "statusText": "OK",
}
`;

exports[`don't mutate PostgrestClient.headers 1`] = `null`;

exports[`missing table 1`] = `
Object {
  "body": null,
  "data": null,
  "error": Object {
    "code": "42P01",
    "details": null,
    "hint": null,
    "message": "relation \\"public.missing_table\\" does not exist",
  },
  "status": 404,
  "statusText": "Not Found",
}
`;

exports[`on_conflict insert 1`] = `
Object {
  "body": Array [
    Object {
      "age_range": "[20,30)",
      "catchphrase": "'fat' 'rat'",
      "data": null,
      "status": "ONLINE",
      "username": "dragarcia",
    },
  ],
  "data": Array [
    Object {
      "age_range": "[20,30)",
      "catchphrase": "'fat' 'rat'",
      "data": null,
      "status": "ONLINE",
      "username": "dragarcia",
    },
  ],
  "error": null,
  "status": 201,
  "statusText": "Created",
}
`;

exports[`stored procedure 1`] = `
Object {
  "body": "ONLINE",
  "data": "ONLINE",
  "error": null,
  "status": 200,
  "statusText": "OK",
}
`;

exports[`switch schema 1`] = `
Object {
  "body": Array [
    Object {
      "age_range": "[1,2)",
      "data": null,
      "status": "ONLINE",
      "username": "supabot",
    },
    Object {
      "age_range": "[25,35)",
      "data": null,
      "status": "OFFLINE",
      "username": "kiwicopple",
    },
    Object {
      "age_range": "[25,35)",
      "data": null,
      "status": "ONLINE",
      "username": "awailas",
    },
    Object {
      "age_range": "[20,30)",
      "data": null,
      "status": "ONLINE",
      "username": "dragarcia",
    },
    Object {
      "age_range": "[20,40)",
      "data": null,
      "status": "ONLINE",
      "username": "leroyjenkins",
    },
  ],
  "data": Array [
    Object {
      "age_range": "[1,2)",
      "data": null,
      "status": "ONLINE",
      "username": "supabot",
    },
    Object {
      "age_range": "[25,35)",
      "data": null,
      "status": "OFFLINE",
      "username": "kiwicopple",
    },
    Object {
      "age_range": "[25,35)",
      "data": null,
      "status": "ONLINE",
      "username": "awailas",
    },
    Object {
      "age_range": "[20,30)",
      "data": null,
      "status": "ONLINE",
      "username": "dragarcia",
    },
    Object {
      "age_range": "[20,40)",
      "data": null,
      "status": "ONLINE",
      "username": "leroyjenkins",
    },
  ],
  "error": null,
  "status": 200,
  "statusText": "OK",
}
`;
